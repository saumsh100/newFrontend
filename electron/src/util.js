
module.exports.getChannelFromVersion = (version) => {
  const splitVersion = version.split('-');
  return splitVersion[1] || null;
};
