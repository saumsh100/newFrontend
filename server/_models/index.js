
const Sequelize = require('sequelize');
const { postgres } = require('../config/globals');

const sequelizeConfig = {
  logging: false,
  dialect: 'postgres',
  host: postgres.host,
  port: postgres.port,
  define: {
    underscored: false,
    paranoid: true,
  },
};

// Don't use NODE_ENV===production
// It doesn't let us debug production locally easily
if (postgres.ssl) {
  sequelizeConfig.dialectOptions = {
    ssl: true,
  };
}

// If true, sequelize will dump all PostgreSQL queries into terminal
if (postgres.logging) {
  sequelizeConfig.logging = console.log // eslint-disable-line
}

// initialize sequelize
const sequelize = new Sequelize(
  postgres.database,
  postgres.username,
  postgres.password,
  sequelizeConfig,
  );

// Test the connection for logs
sequelize
  .authenticate()
  .then(() => {
    console.log(`Sequelize connected to database=${postgres.database} at host:port=${postgres.host}:${postgres.port}`);
  })
  .catch(err => {
    console.error('postgres config', postgres);
    console.error('Unable to connect to the database:', err);
  });

// Import and store all models.
const models = [];
// models.push((require('./Address').default(sequelize, Sequelize)));
models.push((require('./Account').default(sequelize, Sequelize)));
models.push((require('./Appointment').default(sequelize, Sequelize)));
models.push((require('./AuthSession').default(sequelize, Sequelize)));
models.push((require('./Call').default(sequelize, Sequelize)));
models.push((require('./Chair').default(sequelize, Sequelize)));
models.push((require('./Chat').default(sequelize, Sequelize)));
models.push((require('./Enterprise').default(sequelize, Sequelize)));
models.push((require('./Family').default(sequelize, Sequelize)));
models.push((require('./Invite').default(sequelize, Sequelize)));
models.push((require('./Patient').default(sequelize, Sequelize)));
models.push((require('./PatientUser').default(sequelize, Sequelize)));
models.push((require('./Permission').default(sequelize, Sequelize)));
models.push((require('./PinCode').default(sequelize, Sequelize)));
models.push((require('./Practitioner').default(sequelize, Sequelize)));
models.push((require('./Practitioner_Service').default(sequelize, Sequelize)));
models.push((require('./PractitionerRecurringTimeOff').default(sequelize, Sequelize)));
models.push((require('./Recall').default(sequelize, Sequelize)));
models.push((require('./Reminder').default(sequelize, Sequelize)));
models.push((require('./Request').default(sequelize, Sequelize)));
models.push((require('./Segment').default(sequelize, Sequelize)));
models.push((require('./SentRecall').default(sequelize, Sequelize)));
models.push((require('./SentReminder').default(sequelize, Sequelize)));
models.push((require('./Service').default(sequelize, Sequelize)));
models.push((require('./SyncClientError').default(sequelize, Sequelize)));
models.push((require('./SyncClientVersion').default(sequelize, Sequelize)));
models.push((require('./TextMessage').default(sequelize, Sequelize)));
models.push((require('./Token').default(sequelize, Sequelize)));
models.push((require('./User').default(sequelize, Sequelize)));
models.push((require('./WaitSpot').default(sequelize, Sequelize)));
models.push((require('./WeeklySchedule').default(sequelize, Sequelize)));

const db = {};
models.forEach((model) => {
  db[model.name] = model;
});

Object.keys(db).forEach((modelName) => {
  if (db[modelName].associate) {
    db[modelName].associate(db);
  }
});

db.sequelize = sequelize;
db.Sequelize = Sequelize;

module.exports = db;
