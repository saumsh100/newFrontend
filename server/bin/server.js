
const http = require('http');
const globals = require('../config/globals');
const app = require('./app');
const Availability = require('../models/Availability');
const Patient = require('../models/Patient');

const server = http.createServer(app);

const io = require('socket.io')(server);

// TODO: implement Thinky!!!
io.on('connection', (socket) => {
  console.log('a new socket connected');
  socket.on('fetchAvailabilities', () => {
    Availability.run()
      .then((availabilities) => {
        socket.emit('receiveAvailabilities', availabilities);
      });
  });
  
  socket.on('fetchPatients', () => {
    Patient.run()
      .then((patients) => {
        socket.emit('receivePatients', patients);
      });
  });
  
  socket.on('addAvailability', (data) => {
    new Availability(data).save().then((availability) => {
      io.sockets.emit('availabilityAdded', availability);
    });
  });
  
  socket.on('removeAvailability', (data) => {
    Availability.get(data.id).then((availability) => {
      availability.delete()
        .then((result) => {
          io.sockets.emit('availabilityRemoved', result);
        });
    });
  });
  
  socket.on('syncClientTrigger', (data) => {
    console.log('sync client triggers!', 'data:', data);
  });
});

server.listen(globals.port, () => {
  console.log(`CareCru HTTP Server is running on port ${globals.port}`);
});
