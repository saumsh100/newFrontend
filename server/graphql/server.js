
import { Router } from 'express';
import graphQLHTTP from 'express-graphql';
import { sequelizeAuthMiddleware } from '../middleware/auth';
import schema from './data/schema';

const GraphQLRouter = Router();

const PORT = 4000;

const GraphQLServer = graphQLHTTP({
  schema,
  graphiql: true,
  pretty: true,
  subscriptionsEndpoint: `ws://localhost:${PORT}/subscriptions`,
});

// Reusing auth middleware as the server is exposed the same way an API would.
GraphQLRouter.all('*', sequelizeAuthMiddleware, GraphQLServer);

export default GraphQLRouter;

