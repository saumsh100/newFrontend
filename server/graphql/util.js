
import { getOffsetWithDefault, offsetToCursor } from 'graphql-relay';

/**
 * A simple function that accepts an array and connection arguments, and returns
 * a connection object for use in GraphQL
 *
 * This is a fork of connectionFromArraySlice
 * from https://github.com/graphql/graphql-relay-js/blob/master/src/connection/arrayconnection.js#L65
 *
 * Diferente form the original, is this case we build the cursor and pageInfo based on args
 * whitout slicing the array.
 * That way we can pass the already sliced or filtered array to the function.
 *
 * For now, this only works with forward pagination.
 */
// eslint-disable-next-line import/prefer-default-export
export function connectionFromArrayWithoutSlice(array, args, meta) {
  const { after, first, last } = args;
  const { arrayLength } = meta;

  if (typeof first === 'number') {
    if (first < 0) {
      throw new Error('Argument "first" must be a non-negative integer');
    }
  }
  if (typeof last === 'number') {
    if (last < 0) {
      throw new Error('Argument "last" must be a non-negative integer');
    }
  }

  const afterOffset = getOffsetWithDefault(after, 0);

  const startOffset = Math.max(afterOffset, -1) + 1;
  const endOffset = startOffset + first;

  const edges = array.map((value, index) => ({
    cursor: offsetToCursor(startOffset + index),
    node: value,
  }));

  const firstEdge = edges[0];
  const lastEdge = edges[edges.length - 1];
  const lowerBound = after ? afterOffset + 1 : 0;
  const upperBound = arrayLength;

  return {
    edges,
    pageInfo: {
      startCursor: firstEdge ? firstEdge.cursor : null,
      endCursor: lastEdge ? lastEdge.cursor : null,
      hasPreviousPage: typeof last === 'number' ? startOffset > lowerBound : false,
      hasNextPage: typeof first === 'number' ? endOffset < upperBound : false,
    },
  };
}
