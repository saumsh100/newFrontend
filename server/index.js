'use strict';

const bodyParser = require('body-parser');
const compress = require('compression');
const express = require('express');
const subdomain = require('express-subdomain');
const handleErrors = require('./helpers/handleErrors');
const { buildFolder, mainReactApp, mySubdomain } = require('./config');
const widgetRouter = require('./routes/widget');
const myRouter = require('./routes/my');

/**
 * App Variables
 */

const server = express();
server.disable('x-powered-by');
const port = parseInt(process.env.PORT || '80', 10);

/**
 *  App Configuration
 */
server.use(
  compress({
    level: 9,
    filter(req, res) {
      return /json|text|javascript|css/.test(res.getHeader('Content-Type'));
    },
  }),
);

// Bind middleware to show favicon and set up static routes
server.use('/', express.static(buildFolder));

// Parse URL query strings into the body,
// extended=true means that we use qs as the parser for Object syntax
server.use(
  bodyParser.urlencoded({
    limit: '10mb',
    extended: true,
  }),
);

/**
 * Routes Definitions
 */
const rootRouter = express.Router();

// Bind subdomain capturing
const subdomainRouter = express.Router();
subdomainRouter.use(widgetRouter);
subdomainRouter.use(myRouter);
rootRouter.use(subdomain(mySubdomain, subdomainRouter));

// Widget on the root
rootRouter.use(widgetRouter);
rootRouter.use(myRouter);
/**
 * Server Activation
 */

// All other traffic, just render app
rootRouter.get('(/*)?', async (req, res, next) => {
  res.sendFile(mainReactApp);
});

server.use(rootRouter);

// Catch errors, log and respond to client
server.use(handleErrors);

server.listen(port, () => {
  console.log(`Listening to requests on http://127.0.0.1:${port}`);
});
