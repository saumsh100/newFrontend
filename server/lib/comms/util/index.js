
import groupBy from 'lodash/groupBy';
import map from 'lodash/map';
import toArray from 'lodash/toArray';

const ATTRS = {
  email: 'email',
  sms: 'mobilePhoneNumber',
  phone: 'mobilePhoneNumber',
};

const ERROR_CODES = {
  email: '1100',
  sms: '1200',
  phone: '1300',
};

const PREFERENCE_TYPES = {
  email: 'emailNotifications',
  sms: 'sms',
  phone: 'phone',
};

const NO_PREFERENCE_ERROR_CODES = {
  email: '2100',
  sms: '2200',
  phone: '2300',
};

/**
 * cannotSend
 *
 * @param patient
 * @param primaryType
 * @returns {{errors: Array, success: Array}}
 */
export function cannotSend(patient, primaryType) {
  // If it is undefined return the error code
  if (!patient[ATTRS[primaryType]]) {
    return ERROR_CODES[primaryType];
  }
}

/**
 * noPreference
 *
 * @param patient
 * @param primaryType
 * @returns {{errors: Array, success: Array}}
 */
export function noPreference(patient, primaryType) {
  const preferences = patient.preferences || {};

  // If it is false return errorCode
  if (!preferences[PREFERENCE_TYPES[primaryType]]) {
    return NO_PREFERENCE_ERROR_CODES[primaryType];
  }
}

/**
 * generateOrganizedPatients will accept an array of patients and an
 * array of primaryTypes and will return the errors and successes based
 * on patientData. The point is to organize the failed responses so that
 * certain jobs can save a bulk amount to save time
 *
 * @param patients
 * @param primaryTypes (ie.// ['email', 'sms'])
 * @returns {{errors: Array, success: Array}}
 */
export function generateOrganizedPatients(patients, primaryTypes) {
  const errors = [];
  const success = [];

  for (const patient of patients) {
    primaryTypes.forEach((primaryType) => {
      const errorCode = cannotSend(patient, primaryType) || noPreference(patient, primaryType);
      if (errorCode) {
        errors.push({
          patient,
          errorCode,
          primaryType,
        });
      } else {
        success.push({ primaryType, patient });
      }
    });
  }

  return { errors, success };
}

/**
 * organizeForOutbox is a function that accepts an array and a selectorPredicate
 * and will group the array based on the selectorPredicate, and then map over that array
 * to return outbox items that have primaryTypes grouped together.
 *
 * NOTE: this is primarily a layer ontop of the mapPatientsTo____ calls that will take the flat
 * array with primaryType and bring it back into primaryTypes
 *
 * @param array (ie.// [{ patient, appointment, primaryType: 'sms' }, { patient, appointment, primaryType: 'email' }])
 * @param selectorPredicate (ie.// ({ patient, appointment }) => `${patient.id}_${appointment.id}`)
 * @param mergePredicate
 * @returns [organizedArray] (ie.// [{ patient, appointment, primaryTypes: ['sms', 'email'] }])
 */
export function organizeForOutbox(array, selectorPredicate, mergePredicate) {
  const groupedMap = groupBy(array, selectorPredicate);
  return toArray(map(groupedMap, mergePredicate));
}
