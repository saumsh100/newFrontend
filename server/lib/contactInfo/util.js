
const orderBy = require('lodash/orderBy');

/**
 * isHead checks if a given patient is the head of the given family by comparing accountId,
 * familyId and headId.
 * @param family
 * @param patient
 * @return {boolean}
 */
const isHead = (family, patient) =>
  !!family &&
  !!patient &&
  family.accountId === patient.accountId &&
  family.id === patient.familyId &&
  family.headId === patient.id;

/**
 * isActive checks if a patient has status equals to the string 'Active'
 * @param patient
 * @return {boolean}
 */
const isActive = ({ status }) => status === 'Active';

/**
 * getFromNewestCreatedFamily orders the a list of patients by its family creation date and
 * grabs the first.
 * @param patients
 */
const getFromNewestCreatedFamily = patients =>
  orderBy(
    patients,
    ({ family: { pmsCreatedAt, createdAt } }) => pmsCreatedAt || createdAt,
    'desc',
  )[0];

/**
 * getOldest orders a list of patients by their birth dates and grabs the first, thus the oldest.
 * @param patients
 */
const getOldest = patients => orderBy(patients, p => p.birthDate, 'asc')[0];

/**
 * getPatientAge is a utility function to extract the patient age form its birth date by year
 * diffing. This was created using plain JS so we can reduce the usage of moment.js.
 * @param patient
 * @return {number}
 */
const getPatientAge = (patient) => {
  const ageDate = new Date(Date.now() - new Date(patient.birthDate).getTime());
  return Math.abs(ageDate.getUTCFullYear() - 1970);
};

/**
 * adultChecker is a configurable function builder that creates a function that will check if an age
 * of a patient is within the adult range provided to it.
 * @param min
 * @param max
 * @return {function(*=): boolean}
 */
const adultChecker = (min, max) => patient =>
  min < getPatientAge(patient) && getPatientAge(patient) < max;

/**
 * grabPoC is a function that takes a patient list and returns a function that filters this list by
 * the given predicate, in case of the filter returns more than one result the fallbackHandler is
 * applied to the list.
 * This function use partially applied arguments so we can reuse the same list of patients but with
 * different predicates and fallbackHandlers.
 * @param patients
 * @return {Function}
 */
const grabPoC = patients => (filterPredicate, fallbackHandler) => {
  const filteredPatients = patients.filter(filterPredicate);
  if (filteredPatients.length === 1) {
    return filteredPatients[0];
  } else if (filteredPatients.length > 1) {
    return fallbackHandler(filteredPatients);
  }

  return undefined;
};

module.exports = {
  adultChecker,
  grabPoC,
  isHead,
  isActive,
  getFromNewestCreatedFamily,
  getOldest,
};
