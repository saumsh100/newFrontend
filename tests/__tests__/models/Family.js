
import { Account, Family, Patient } from '../../../server/_models';
import { omitProperties }  from '../../util/selectors';

async function wipeFamilyTable() {
  await Family.destroy({
    where: {},
    force: true,
  });
}

async function wipePatientTable() {
  await Patient.destroy({
    where: {},
    force: true,
  });
}

async function wipeAccountTable() {
  await Account.destroy({
    where: {},
    force: true,
  });
}

const accountId = 'e13151a6-091e-43db-8856-7e547c171754';
const makeData = (data = {}) => (Object.assign({
  accountId,
}, data));

const enterpriseId = 'ef3c578f-c228-4a25-8388-90ee9a0c9eb4';
const makeAccountData = (data = {}) => (Object.assign({
  id: accountId,
  name: 'Test Account',
  enterpriseId,
}, data));

const makePatientData = (data = {}) => Object.assign({
  accountId,
  firstName: 'Test',
  lastName: 'Patient',
}, data);

const fakeAccountId = 'f23151a6-091e-43db-8856-7e547c171754';
const fail = 'Your code should be failing but it is passing';

describe('models/Family', () => {
  beforeEach(async () => {
    await wipeFamilyTable();
    await wipePatientTable();
    await wipeAccountTable();
  });

  afterAll(async () => {
    await wipeFamilyTable();
    await wipePatientTable();
    await wipeAccountTable();
  });

  describe('Data Validation', () => {
    test('should be able to save a Family without id provided', async () => {
      const data = makeData();
      await Account.create(makeAccountData());
      const family = await Family.create(data);
      expect(omitProperties(family.dataValues, ['id'])).toMatchSnapshot();
    });

    test('should have null value for pmsId', async () => {
      const data = makeData();
      await Account.create(makeAccountData());
      const family = await Family.create(data);
      expect(family.pmsId).toBe(null);
    });

    test('should throw error for no accountId provided', async () => {
      const data = makeData({ accountId: undefined });
      await Account.create(makeAccountData());
      try {
        await Family.create(data);
        throw new Error(fail);
      } catch (err) {
        expect(err.name).toEqual('SequelizeValidationError');
      }
    });

    test('should fail if accountId does not reference an existing account', async () => {
      const data = makeData({ accountId: fakeAccountId });
      try {
        await Family.create(data);
        throw new Error(fail);
      } catch (err) {
        expect(err.name).toEqual('SequelizeForeignKeyConstraintError');
      }
    });
  });

  describe('Relations', () => {
    beforeEach(async () => {
      await wipeAccountTable();
    });

    test('should be able to fetch account relationship', async () =>  {
      await Account.create(makeAccountData());
      const { id } = await Family.create(makeData());

      const family = await Family.findOne({
        where: { id },
        include: [
          {
            model: Account,
            as: 'account',
          },
        ],
      });

      expect(family.accountId).toBe(family.account.id);
    });

    test('should not fail if you are trying to join a chair if chairId is null', async () =>  {
      await Account.create(makeAccountData());
      const { id } = await Family.create(makeData());
      await Patient.bulkCreate([
        makePatientData({ familyId: id }),
        makePatientData({ familyId: id }),
      ]);

      const f = await Family.findOne({
        where: { id },
        include: [
          {
            model: Patient,
            as: 'patients',
          },
        ],
      });

      expect(f.patients.length).toBe(2);
    });
  });
});
