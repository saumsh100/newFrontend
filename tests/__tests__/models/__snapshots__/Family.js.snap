// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models/Family Data Validation should be able to save a Family without id provided 1`] = `
Object {
  "accountId": "62954241-3652-4792-bae5-5bfed53d37b7",
  "headId": null,
  "pmsId": null,
}
`;
